File: CYBaseProject/CYBaseProject/Code/Utils/UI/UIFont+CYUtils.m
//
//  UIFont+CYUtils.m
//  CYBaseProject
//
//  Created by HuangQiSheng on 4/21/16.
//  Copyright © 2016 Jasper. All rights reserved.
//

#import "UIFont+CYUtils.h"

#import "UIScreen+CYUtils.h"

@implementation UIFont (CYUtils)

+ (CGFloat)screenAdaptFactor {
    
    static CGFloat standardWidth = 320.f;
    static CGFloat factor = 0;
    
    if (factor == 0) {
        
        factor = [UIScreen mainScreen].bounds.size.width / standardWidth;
    }
    return factor;
}

+ (instancetype)screenAdapt6PlusOnlySystemFontOfSize:(CGFloat)fontSize {
    
    return [self screenAdaptSystemFontOfSize:fontSize
                              adapt6PlusOnly:YES];
}

+ (instancetype)screenAdaptSystemFontOfSize:(CGFloat)fontSize
                             adapt6PlusOnly:(BOOL)adapt6PlusOnly {
    
    if (!adapt6PlusOnly
        || [UIScreen mainScreen].is6PlusScreen) {
        
        return [UIFont systemFontOfSize:(fontSize * [self screenAdaptFactor])];
    } else {
        
        return [UIFont systemFontOfSize:fontSize];
    }
}

+ (instancetype)screenAdapt6PlusOnlyBoldSystemFontOfSize:(CGFloat)fontSize {
    
    return [self screenAdaptBoldSystemFontOfSize:fontSize
                                  adapt6PlusOnly:YES];
}

+ (instancetype)screenAdaptBoldSystemFontOfSize:(CGFloat)fontSize
                                 adapt6PlusOnly:(BOOL)adapt6PlusOnly {
    
    if (!adapt6PlusOnly
        || [UIScreen mainScreen].is6PlusScreen) {
        
        return [UIFont boldSystemFontOfSize:(fontSize * [self screenAdaptFactor])];
    } else {
        
        return [UIFont boldSystemFontOfSize:fontSize];
    }
}

+ (CGFloat)screenAdaptFontSizeWith:(CGFloat)fontSize {
    
    return [self screenAdaptFontSizeWith:fontSize
                          adapt6PlusOnly:NO];
}

+ (CGFloat)screenAdapt6PlustOnlyFontSizeWith:(CGFloat)fontSize {
    
    return [self screenAdaptFontSizeWith:fontSize
                          adapt6PlusOnly:YES];
}

+ (CGFloat)screenAdaptFontSizeWith:(CGFloat)fontSize
                    adapt6PlusOnly:(BOOL)adapt6PlusOnly {
    
    if (!adapt6PlusOnly
        || [UIScreen mainScreen].is6PlusScreen) {
        
        return (fontSize * [self screenAdaptFactor]);
    } else {
        
        return fontSize;
    }
}

@end

File: CYBaseProjectSwift/CYBaseProjectSwift/Code/Utils/UI/UIFont+CYUtils.swift
//
//  UIFont+CYUtils.swift
//  CYBaseProjectSwift
//
//  Created by HuangQiSheng on 4/22/16.
//  Copyright © 2016 Jasper. All rights reserved.
//

import UIKit

extension UIFont {
    
    static var screenFactor: CGFloat {
        
        get {
            
            return UIScreen.mainScreen().bounds.size.width / 320.0
        }
    }
    
    class func systemFontOfSize(fontSize: CGFloat, adaptScreen: Bool) -> UIFont {
        
        return self.systemFontOfSize(self.fontSize(fontSize, adaptScreen: adaptScreen))
    }
    
    class func systemFontOfSize(fontSize: CGFloat, adaptFontFor6Plus: Bool) -> UIFont {
        
        return systemFontOfSize(fontSize,
                                adaptScreen: (adaptFontFor6Plus && UIScreen.mainScreen().is6PlusScreen))
    }
    
    class func boldSystemFontOfSize(fontSize: CGFloat, adaptScreen: Bool) -> UIFont {
        
        return self.boldSystemFontOfSize(self.fontSize(fontSize, adaptScreen: adaptScreen))
    }
    
    class func boldSystemFontOfSize(fontSize: CGFloat, adaptFontFor6Plus: Bool) -> UIFont {
        
        return boldSystemFontOfSize(fontSize,
                                    adaptScreen: (adaptFontFor6Plus && UIScreen.mainScreen().is6PlusScreen))
    }
    
    class func fontSize(fontSize: CGFloat, adaptScreen: Bool) -> CGFloat {
        
        if adaptScreen {
            
            return fontSize * self.screenFactor
        } else {
            
            return fontSize
        }
    }
    
    class func fontSize(fontSize: CGFloat, adaptFontFor6Plus:Bool) -> CGFloat {
        
        return self.fontSize(fontSize,
                             adaptScreen: (adaptFontFor6Plus && UIScreen.mainScreen().is6PlusScreen))
    }
}
